### ISSUE TITLE
Data-loss bugs with priority

### ISSUE THREAD

**Issue by skoczen** [firebase/angularfire#621]
Hey Firebase folks,

Running into two bugs here, one that’s causing data loss.

I’ve lost about two days to this, but in the end, have a clean codepen that demonstrates it.  Hooray!

http://codepen.io/skoczen/pen/eNdJqd

Here's the bugs: 

**Bug 1**. The list on the left isn’t ordered by priority, even though priorities are set, and they’re sorted correctly in forge.  This is what caused my original exploration, and appears to be a bug somewhere - or, angularfire doesn't work the way the docs lead me to believe.

You can see this one by just loading the pen, and clicking reset data.  The list on the left *should* be sorted by priority, but isn't.

**Bug 2**:  This is the scary one.  The instructions to replicate are in the codepen on the right, but in short, setting data with a priority, then setting another node in the same ref in a later request strips all the priority data off unrelated nodes! 

Happy to do whatever I can to debug and help out on these - our development is totally blocked on them.

Thanks!
-Steven

**Comment by jamestalmage** [comment]
@skoczen 
I can't reproduce the data loss you are describing, but I can help you with your sorting problem.

The issue is that you are using `orderByPriority` incorrectly. That returns a new ref that points to sorted data, while the initial ref remains unchanged. You can actually think of refs as immutable data structures, even though the data they reference very mutable. A ref points to a storage location, with a prescribed sorting method for it's children. Those values (referenced location and sorting) can not be changed. Instead you can create a new reference that points to a new location using `.child()`, or a new reference that prescribes a new sorting strategy `orderByX()`. So `ref.child('a').child('b').orderByPriority()` actually creates 3 new refs, we just discard the middle two.

In your code you do the following:
```javascript
ref.orderByPriority();
```

But you never assign the result of `orderByPriority()` to anything, so you don't have access to the sorted data. What you need to do is this:

```javascript
var unsorted = new Firebase('myurl');
var sorted = unsorted.orderByPriority();
```

This is definitely not the first time I have seen people confused by this. Perhaps the docs could cover it better, but I will leave that to the Firebase team.

Unfortunately, your issues don't end there. `orderByPriority` will only affect the ordering of that nodes direct children (*not it's grandchildren, great-grandchildren, etc*). So calling `orderByPriority` where you did will only affect the ordering of that nodes two children (`news_feed` and `hearts` in this case), while the sorting of the deeper nodes remains unaffected. 

Also, anytime you are doing an ng-repeat, you probably want to consider using [$firebaseArray](https://www.firebase.com/docs/web/libraries/angular/guide/synchronized-arrays.html)

So, putting that all together, this is probably more like what you want:

```javascript
var ref = new Firebase("https://buddyup-bug-explorer.firebaseio.com/users/1");
$scope.me.newsfeed = $firebaseArray(ref.child('news_feed').orderByPriority());
```

You would also need to update your template to point the `ng-repeat` directive at `me.newsfeed` instead of `me.data`. I have [updated your codepen](http://codepen.io/jamestalmage/pen/KpgzBW) with a working example.

Also, it is hard to tell where you are headed with such a simple example, but I would advise reading the docs on [structuring your data](https://www.firebase.com/docs/web/guide/structuring-data.html#section-flatten), as well as the blog post on [denormalizing your data](https://www.firebase.com/blog/2013-04-12-denormalizing-is-normal.html). They are hard concepts, but critically important.

**Comment by skoczen** [comment]
Hey @jamestalmage, 

Most important stuff first:

## Bug 2
I'm able to replicate bug #2 by following the instructions on the right, using a fresh, incognito browser.   It happens 100% of the time.  What results are you seeing?


## Bug 1 (aka confusion)
Thanks for all the info on the sorting. I'd completely misunderstood the `orderByPriority`, thinking that it sorted in-place.  Makes more sense this way. 

I've read those docs extensively, and we've got pretty heavily denormalized data structure - the challenge has been getting sorting and updating to work in a heavily race-conditioned setup (including clients and servers racing).  As such, I've shied away from `$firebaseArray`, as the docs are pretty strongly against using anything that's array-like in your setup.

From what I'm seeing in your details above though, using `$firebaseArray` doesn't take us down the dreaded firebase array road, but is instead a convenience wrapper for the front end. That makes a bunch more sense.

One thing I'm trying to sort through on these refs is if (and at what number) there's a performance penalty for opening up too many.  Any insights?

Thanks much!

**Comment by katowulf** [comment]
The cool part about $firebaseArray is that it works as an object with the server, using push ids, and as an array locally, giving you the best of both worlds.

Note also that if you're just getting started, I'd discourage using priorities and work with the new query tools, based on orderByChild(), which work of fields in the record. Those are where we are headed in the future.

**Comment by skoczen** [comment]
@katowulf - got it.  About four months in, so would prefer not to do massive changes. Is priority going away though?

@jamestalmage, here's an animated gif of the data-loss bug:

![bug](https://cloud.githubusercontent.com/assets/200635/7736341/b8db47cc-ff08-11e4-866e-67e81874862c.gif)

**Comment by katowulf** [comment]
Priority will continue to be supported for existing versions of the SDK. I'm guessing it will disappear in the next major iteration of the API--possible 3.x. That's purely speculation as none of this has been hashed out yet and it's a ways down the road map. However, consider moving to orderByChild() where possible as it has more robust capabilities and will be compatible with more features in the future.

**Comment by katowulf** [comment]
Please start by updating to AngularFire 1.1.1 as there was a bug in the loading promise, which would also affect your $bindTo promise.

Internally, and as documented, AngularFire refers to priorities using $priority, not .priority. It will tolerate `.priority` in data, but this does not extend to the `$bindTo` function, due to some limitations in how `$parse` and `$watch` work. Try setting $priority in $scope.me.data instead. That should help.

**Comment by jamestalmage** [comment]
@skoczen 
Whoa, That just does not happen for me. The priorities stay put (even in your codepen).

Did you try my updated one? It's using newer versions of every library. Either way, it is still weird (I do not have the problem even using the older dependencies). 

What browser are you using? Looks like OSX Chrome to me (which is what I tested on). Maybe try in Incognito mode so browser extensions are not in play (long shot).

**Comment by skoczen** [comment]
@katowulf

Thanks for the info on deprecation - will see how much work it'd take to move all of our stuff over to `orderByChild`.  Losing the ordering in forge is a bummer though, especially for debugging - and the fact that we can explicitly order every node, and then not have to worry about missing `orderBy`s on all our platforms  (about 70% of the data is injected by the server, not clients) is a big bonus. 

Granted that this is all subject to change, would you guess that the version that drops priority would add the ability to set a default ordering key for the entire datastore?

Gotcha on `$priority` - we actually set them with `setPriority`, and only use `.priority` for writes via the REST api, but was trying to get the example all in one pen, so used `.priority` here.  Good to know it matters.

@jamestalmage 
My pen has been updated to 1.1.1. No difference.

Here's chrome incognito (gif for completeness),
![icognito](https://cloud.githubusercontent.com/assets/200635/7738767/c7e5df38-ff1d-11e4-9ac8-e1ffc8f30bc9.gif)


And Safari.  Latest versions of both.
![safari](https://cloud.githubusercontent.com/assets/200635/7738748/85fc53cc-ff1d-11e4-8b74-d78b73caf6e0.gif)

Firefox just in case it was webkit.
![firefox](https://cloud.githubusercontent.com/assets/200635/7738806/2cb67fd0-ff1e-11e4-9501-bda5b6ce0cb2.gif)


Really, really weird that it's not showing for you.  I see this on versions from 0.9.x through to 1.1.1.

I've also seen it over the past two days, trying to work out why all the priorities were being stripped off our data, how, and when. It's very, very reproducible here, and really consistent in how it behaves.

What can I do to help repo this for you and get it solved?

**Comment by skoczen** [comment]
@katowulf as a follow up, on the trail of your `.priority` vs `$priority`, the original case that triggered finding this bug:

1. Nodes with `.priority` are set via the REST api (since it's the only way I can find to set priority via python)
2. Data is saved in the same ref tree, but a different branch via the client. (Bound by object via `.$bindTo`)
3. `.priority`s for all data in the tree are killed.

If `.$bindTo` ignores all `.priority` values, and `$bindTo` just writes an entire tree, instead of checking for node changes, this could cause the problem I'm describing.  It also means `.$bindTo` is frighteningly, frighteningly dangerous, and should *never* be used on any data that contains priorities.  It also (to me) feels like a bug - using `.$bindTo`, to me, shouldn't delete any data that was created elsewhere.

Thoughts?

**Comment by katowulf** [comment]
You're never going to have good luck trying to maintain priorities on nested data. $bindTo doesn't have any way to know that there are priorities on nested child nodes. It is discouraged for any sort of nested collections in the docs, as is $firebaseObject. Nested priorities on child nodes are not going to be maintained there any more than they would if you called set() on on object which has nested child nodes (which is exactly what happens internally in $firebaseObject). 

By the way, this discussion is one primary and motivating factor for switching to orderByChild() and dropping priority usage.

Ideally, if you want to manage a sorted collection, you should use $firebaseArray, which is designed for this. You also shouldn't nest your lists under your user object. As they grow, you're going to see performance degrade in fetching user profiles or iterating users in a list.

```
// preferably, move this to its own node in Firebase instead of nesting it in the user profile
var ref = new Firebase("https://buddyup-bug-explorer.firebaseio.com/users/1/news_feed");
var feeds = $firebaseArray(ref);
```

Regarding the question of ordering in your dashboard, I'd go ahead and continue to set priorities in addition to setting child values for the time being, until the dashboard allows you to sort these in a more meaningful way.

**Comment by jamestalmage** [comment]
OK.
I finally reproduced, I did not follow the directions (I skipped step 2).

Looking into it now.

**Comment by jamestalmage** [comment]
Not fixed, but I have narrowed it down a bit further.

See: http://codepen.io/jamestalmage/pen/mJrOWo

Instead of resetting data by setting a value on scope, I reset by calling `ref.set` directly.

You now no longer need step 2.

Basically `$firebaseObject` does not read priorities for child values, but it will write priorities back to firebase if you set `.priority` on the scope (but not `$priority`). This is a weird and inconsistent behavior that should probably be addressed.

Your options for now are:
 * Stop using priority and use a named child value and `orderByChild` queries.
 * Denormalize your data so you are not trying to bind scope to a node with children that have priority values.
 * Do not bind to that parent node, but bind to all the child values individually (using `$firebaseArray` and `$firebaseObject` as appropriate). This is probably not recommended unless you know the total data stored under the entire tree is very small (prefer denormalization).

**Comment by skoczen** [comment]
Hey @jamestalmage  and @katowulf ,

## Schema & denormalized data

So in regards to denormalized data, I'd love to get some clarity as to your meanings.

From my understanding, denormalizing means keeping a copy of the data in the most user-centric location, i.e. to limit total accesses needed for the data, instead minimizing total copies on the back-end (DRY).  The issue of nesting data seems separate. (But, as we're finding on this bug, relevant.)

In this case, here's my data tree, which seems reasonably denormalized to me:


```
SCHEMA A
- user
    - 1  (a user's data, read and write-limited to that user only, unless otherwise specified)
        - public_profile (object, read-access to other users)
        - profile (object)
        - news_feed (object, only relevant to the user last 100 used in template)
        - buddies (object, list of ids)
        - classes (list of ids)
        - etc (similar)
- class
    - 1
        - profile
```

From what I can see, this is cleanly denormalized data.  Identical stories will appear in many news feeds, and users keep local lists of their classes, buddies, and similar data.  In no cases would we ever be connecting to or iterating the user or class nodes.

The goal was to reduce connections and keep the tree small - so that a single user could just bind to their user node, and be done.  In practice, this worked well, except for the deletion of `.priority` nodes.

@jamestalmage , your third option suggests that instead of: 

```js
var userRef = new Firebase("https://" + window.firebase_endpoint + "/users/" + $scope.me.uid + "/");
$firebaseObject(userRef).$bindTo($scope, "me");
```

I should:

```js
var profileRef = new Firebase("https://" + window.firebase_endpoint + "/users/" + $scope.me.uid + "/");
$firebaseObject(profileRef).$bindTo($scope, "me.profile");

var newsRef = new Firebase("https://" + window.firebase_endpoint +  "/users/" + $scope.me.uid + "/news_feed/");
$scope.me.news_feed = $firebaseArray(newsRef.limitToFirst(100));

// etc
```

This is certainly doable, as long as keeping ~20 open refs doesn't create a performance problem (I could find no docs on this.)


However, I'm also hearing you use "denormalize" to mean "flat" - which suggests that this would be a better structure.

```
SCHEMA B
-user_1_public_profile (object, read-access to other users)
-user_1_profile (object)
-user_1_news_feed (object, only relevant to the user last 100 used in template)
-user_1_buddies (object, list of ids)
-user_1_classes (list of ids)
-user_1_etc (similar)
-class_1_profile
```

In this case, the bindings would be similar to my second setup, but the security rules much hairier.  The number of connections seems constant.

So, a couple of questions on best-practices:

1.  Which of these two schemas would you recommmend, and why?
2.  Is there a maximum number of open references we should make sure we don't exceed? We'll be largely on mobile, expecting 3g.
3.  What is the value range for "very small" in your "total data stored under the tree is very small"?


## The `.priority` bug

Agreed it should be addressed somehow - or at the least, a giant red warning should live with the `$bindTo` documents noting that it will delete all `.priority` values on all child nodes with any save, and should never be used with `.priority`.

FWIW, I'd much prefer that the problem be addressed - having simple three-way data binding where we didn't have to worry about any of that stuff was one of the big selling points on firebase for us.  Since we're already invested, we'll clearly work around, but I'd have looked harder if I knew that things like this were true for `$bindTo`.


Finally, thanks for your quick responses, and hashing through all of this.  It's much appreciated.

**Comment by jamestalmage** [comment]
~20 open refs will be no problem. The docs make reference to being able to add "thousands of listeners", and listeners are what matter in terms of bandwidth. (aside: Creating a ref does not cause any data to be transferred. Calling `ref.set(..)` or `ref.on(..)` does.)

As for denormalization, your original schema looks fine, with the only point of concern being the newsfeed. Could you have a separate tree for newsfeed items and use a `limitTo(100)` query? Basically anytime you are thinking `array` or `list`, it's a candidate for denormalization.

One other note. You probably should not be using numeric id's for users/classes. Instead use `ref.push` to create a random id for you.

**Comment by skoczen** [comment]
@jamestalmage thanks for the response - couple questions/clarifications.

News feed: what's the advantage of 

```
-user

**Comment by skoczen** [comment]
Update: I think I see what your concern was, re: `news_feed`, as it's spec'd out above - that I'd be binding to the parent node, which would contain the `news_feed` list.  The concern makes sense in that case, and it's not clear from my example that we're not doing it.

If I've read that right, all good!

**Comment by katowulf** [comment]
> Agreed it should be addressed somehow - or at the least, a giant red warning should live with the $bindTo documents noting that it will delete all .priority values on all child nodes with any save, and should never be used with .priority.

This isn't a limitation of $bindTo. The same thing will happen with any $firebaseObject when you call $save(). There is already a big yellow warning in $bindTo that tells you to be careful using this with nested data structures, a warning in the Purpose of AngularFire section that says to avoid nested collections, and a note in the description of $firebaseObject that says it is not intended for collections.

@jamestalmage don't spend a lot of time working up a PR to try and fix this. I don't want to deal with the complexities of nested collections and child priorities--that's a deep rabbit hole of trouble. Since priorities are no longer the best way to handle sorting, I'm going to close this as won't fix. 

The working solution here is to avoid priorities where possible, reference collections directly instead of nesting them in synchronized objects, or to ignore me and use $extend to change the behavior to whatever you want.

**Comment by skoczen** [comment]
@katowulf and @jamestalmage 

Ok then.  FWIW, the official firebase docs (where I've spent most of my time) don't mention anything about the deprecation of priority, and the behavior in that codepen took me completely by surprise.

The [angularfire documentation on $bindTo](https://www.firebase.com/docs/web/libraries/angular/api.html#angularfire-firebaseobject-bindtoscope-varname) also doesn't have a yellow box, or *any* mention that `.priority` will be deleted, or anything about nesting.

I do see a yellow box on the [step-by-step-guild](https://www.firebase.com/docs/web/libraries/angular/guide/synchronized-objects.html), but it only mentions performance, not **silent deletion of data**.

Maybe you're referring to some different documentation out there that I don't know about?

Maybe I'm being totally unreasonable, but if a part of the system doesn't behave consistently and there's a potential for data loss, there should be warnings everywhere possible. The firebase client warns if a `orderBy` is being performed on a child without an index. A similar warning on nodes with nested priorities seems reasonable to me. Especially since that data will be silently deleted on any write.

I lost two days to this bug, and read every single one of the pages you've linked multiple times before creating the pen and posting this issue.  Maybe I'm just a terrible reader, but it's possible that some combination of documentation, warnings, or bugfixes could make this easier on the next person.

Your call, of course.

Thanks again for the quick and detailed responses.

**Comment by katowulf** [comment]
I understand that you had a bad experience with this and I appreciate the effort you've put into describing the problem and troubleshooting it here. It truly is encouraging to see someone going to great lengths to troubleshoot their own issues. Your patience is also commendable. I'll add a task to review the docs and see if the messaging around nested collections can be improved.

Note that "won't fix" does not mean "you were wrong" or in any way that I don't think the behavior here could be more desirable. It means that this is a difficult topic with a great deal of complexity and I'm choosing to punt on it because I don't see the benefit as being worth the cost. Mostly because there are readily available, simple answers to this:  referencing the collection directly in your Firebase ref (avoiding nesting), using $firebaseArray for collections, and using a child field instead of a priority.

**Comment by skoczen** [comment]
Hey @katowulf ,

I totally dig the complexity, and I've punted on many similar things in my career, for similar reasons.  I did it on [will](https://github.com/skoczen/will) just today.  I get it.

I think there's a big road hazard here that includes data loss, and it's not adequately marked.  You both were great in understanding and communicating the scope and cause of the problem - but without your help, I'd just have concluded that firebase was broken, unreliable, and probably unhappily shifted us to different technology.

Glad to hear you're adding a task to update the docs.  In it, I'd encourage a close read-through of the two docs I linked to (the only ones I can find from the firebase site.)  

I can't find anything in those docs that state that priority values will be deleted without notice when using `$firebaseObject`, or an example of how to re-structure data that would otherwise fall into that pattern.  Both would be helpful, to me.

I've said my piece on this one, definitely do what you think is best.

Thanks again for the responsiveness and clarity.
